### ISSUE TITLE
Buildkit on coreOS failed with xattr

### ISSUE THREAD

**Issue by WTFKr0** [openfaas/openfaas-cloud#312]
Build fail when buildkit is running on a CoreOS Host

## Expected Behaviour
Build passed

## Current Behaviour
Got an error on the COPY stage of the go template
`

## Possible Solution
Maybe related to https://github.com/genuinetools/img/issues/45
Or a buildkit issue itself

## Steps to Reproduce (for bugs)
Do a build on a builkit container running on a coreos Host

## Your Environment
* Docker version `docker version` (e.g. Docker 17.0.05 ):
18.03.1-ce

* Are you using Docker Swarm or Kubernetes (FaaS-netes)?
Docker Swarm

* Operating System and version (e.g. Linux, Windows, MacOS):
Container Linux by CoreOS 1745.7.0 (Rhyolite)
The backend FS is ext4 : 
```
/dev/sda9       ext4       27G  9.7G   16G  38% /
```

* builkit
moby/buildkit:v0.3.1

## Complete logs

* of-builder
```
v: 2018-10-29T08:44:41Z [internal] load build definition from Dockerfile
v: 2018-10-29T08:44:41Z [internal] load build definition from Dockerfile 0.00s
v: 2018-10-29T08:44:41Z [internal] load .dockerignore
v: 2018-10-29T08:44:41Z [internal] load .dockerignore
v: 2018-10-29T08:44:41Z [internal] load .dockerignore 0.00s
v: 2018-10-29T08:44:41Z [internal] load .dockerignore
v: 2018-10-29T08:44:41Z [internal] load build definition from Dockerfile
status: sha256:1fe3318f282332e55350c3f5d3e31966f856272eb3b619eef4c5922f19bbf1a4 transferring context: 0
status: sha256:1fe3318f282332e55350c3f5d3e31966f856272eb3b619eef4c5922f19bbf1a4 transferring context: 0
status: sha256:1fe3318f282332e55350c3f5d3e31966f856272eb3b619eef4c5922f19bbf1a4 transferring context: 2
status: sha256:0123f550ae4c0a04345b683bc463f347dbd4a90e97922f89e66513fd44b55b14 transferring dockerfile: 0
status: sha256:0123f550ae4c0a04345b683bc463f347dbd4a90e97922f89e66513fd44b55b14 transferring dockerfile: 96
status: sha256:0123f550ae4c0a04345b683bc463f347dbd4a90e97922f89e66513fd44b55b14 transferring dockerfile: 1288
v: 2018-10-29T08:44:41Z [internal] load .dockerignore 0.02s
v: 2018-10-29T08:44:41Z [internal] load build definition from Dockerfile 0.04s
v: 2018-10-29T08:44:41Z [internal] load metadata for docker.io/library/alpine:3.8
v: 2018-10-29T08:44:41Z [internal] load metadata for docker.io/library/golang:1.10.4-alpine3.8
v: 2018-10-29T08:44:41Z [internal] load metadata for docker.io/library/golang:1.10.4-alpine3.8 0.96s
v: 2018-10-29T08:44:41Z [internal] load metadata for docker.io/library/alpine:3.8 1.09s
v: 2018-10-29T08:44:42Z [internal] helper image for file operations
v: 2018-10-29T08:44:42Z [stage-1 1/8] FROM docker.io/library/alpine:3.8@sha256:621c2f39f8133acb8e64023a94dbdf0d5ca81896102b9e57c0dc184cadaf5528
v: 2018-10-29T08:44:42Z [stage-1 2/8] RUN apk --no-cache add ca-certificates
v: 2018-10-29T08:44:42Z [stage-1 3/8] RUN addgroup -S app && adduser -S -g app app
v: 2018-10-29T08:44:42Z [stage-1 4/8] RUN mkdir -p /home/app
v: 2018-10-29T08:44:42Z [builder 1/5] FROM docker.io/library/golang:1.10.4-alpine3.8@sha256:55ff778715a4f37ef0f2f7568752c696906f55602be0e052cbe147becb65dca3
v: 2018-10-29T08:44:42Z [builder 2/5] RUN apk --no-cache add curl     && echo Pulling watchdog binary from Github.     && curl -sSL https://github.com/openfaas/faas/releases/download/0.9.6/fwatchdog > /usr/bin/fwatchdog     && chmod +x /usr/bin/fwatchdog     && apk del curl --no-cache
v: 2018-10-29T08:44:42Z [internal] load build context
v: 2018-10-29T08:44:42Z [builder 3/5] COPY . .
v: 2018-10-29T08:44:42Z [builder 4/5] RUN test -z $(gofmt -l $(find . -type f -name '*.go' -not -path ./vendor/* -not -path ./function/vendor/*)) || { echo Run "gofmt -s -w" on your Golang code; exit 1; }
v: 2018-10-29T08:44:42Z [builder 5/5] RUN CGO_ENABLED=0 GOOS=linux     go build --ldflags -s -w -a -installsuffix cgo -o handler . &&     go test $(go list ./... | grep -v /vendor/) -cover
v: 2018-10-29T08:44:42Z [stage-1 5/8] COPY --from=builder /usr/bin/fwatchdog         .
v: 2018-10-29T08:44:42Z [stage-1 6/8] COPY --from=builder /go/src/handler/function/  .
v: 2018-10-29T08:44:42Z [stage-1 7/8] COPY --from=builder /go/src/handler/handler    .
v: 2018-10-29T08:44:42Z [stage-1 8/8] RUN chown -R app /home/app
v: 2018-10-29T08:44:42Z [internal] helper image for file operations 0.00s
v: 2018-10-29T08:44:42Z [stage-1 1/8] FROM docker.io/library/alpine:3.8@sha256:621c2f39f8133acb8e64023a94dbdf0d5ca81896102b9e57c0dc184cadaf5528 0.00s
v: 2018-10-29T08:44:42Z [internal] load build context 0.00s
status: sha256:d2ef066b3c6574aafc9065ff9f1537ddab570a2a24b900437acdc77ef84a7b48 resolve docker.io/tonistiigi/copy:v0.1.7@sha256:9aab7d9ab369c6daf4831bf0653f7592110ab4b7e8a33fee2b9dca546e9d3089 0
status: sha256:6be3389dcb13329826dbd6af9923eb8bdc6910e53a19a8bc33dff54e64577876 resolve docker.io/library/alpine:3.8@sha256:621c2f39f8133acb8e64023a94dbdf0d5ca81896102b9e57c0dc184cadaf5528 0
v: 2018-10-29T08:44:42Z [internal] helper image for file operations 0.00s
v: 2018-10-29T08:44:42Z [builder 1/5] FROM docker.io/library/golang:1.10.4-alpine3.8@sha256:55ff778715a4f37ef0f2f7568752c696906f55602be0e052cbe147becb65dca3 0.00s
status: sha256:2e76a94367c8454673ca68b78c58308a939330a6e375f7c0946f5ce874be91f8 resolve docker.io/library/golang:1.10.4-alpine3.8@sha256:55ff778715a4f37ef0f2f7568752c696906f55602be0e052cbe147becb65dca3 0
v: 2018-10-29T08:44:42Z [internal] load build context
v: 2018-10-29T08:44:42Z [stage-1 2/8] RUN apk --no-cache add ca-certificates 0.00s
v: 2018-10-29T08:44:42Z [stage-1 3/8] RUN addgroup -S app && adduser -S -g app app 0.00s
v: 2018-10-29T08:44:42Z [stage-1 4/8] RUN mkdir -p /home/app 0.00s
v: 2018-10-29T08:44:42Z [builder 2/5] RUN apk --no-cache add curl     && echo Pulling watchdog binary from Github.     && curl -sSL https://github.com/openfaas/faas/releases/download/0.9.6/fwatchdog > /usr/bin/fwatchdog     && chmod +x /usr/bin/fwatchdog     && apk del curl --no-cache
v: 2018-10-29T08:44:42Z [builder 2/5] RUN apk --no-cache add curl     && echo Pulling watchdog binary from Github.     && curl -sSL https://github.com/openfaas/faas/releases/download/0.9.6/fwatchdog > /usr/bin/fwatchdog     && chmod +x /usr/bin/fwatchdog     && apk del curl --no-cache 0.00s
status: sha256:d9ba16af7b0f1c0c7131d3f26a2fe10b7ba8466ba03cc624fefaa95ce3d1c5d7 transferring context: 0
status: sha256:d9ba16af7b0f1c0c7131d3f26a2fe10b7ba8466ba03cc624fefaa95ce3d1c5d7 transferring context: 96
status: sha256:d9ba16af7b0f1c0c7131d3f26a2fe10b7ba8466ba03cc624fefaa95ce3d1c5d7 transferring context: 2915
v: 2018-10-29T08:44:42Z [internal] load build context 0.02s
v: 2018-10-29T08:44:42Z [builder 3/5] COPY . .
log: sha256:5c20ac7ed3bd70180e54ece06c7e24606e7eaf45c7bf33fba39c41176004cc61
panic: failed to copy /src-0/. to go/src/handler/: failed to copy xattrs: failed to set xattr "security.selinux" on go/src/handler/Dockerfile: operation not supported

goroutine 1 [running]:

log: sha256:5c20ac7ed3bd70180e54ece06c7e24606e7eaf45c7bf33fba39c41176004cc61
main.main
log: sha256:5c20ac7ed3bd70180e54ece06c7e24606e7eaf45c7bf33fba39c41176004cc61
()
	/go/src/github.com/tonistiigi/copy/cmd/copy/main.go:57
log: sha256:5c20ac7ed3bd70180e54ece06c7e24606e7eaf45c7bf33fba39c41176004cc61
 +0x2af

v: 2018-10-29T08:44:42Z [builder 3/5] COPY . . 0.16s
```

* builkit
```
time="2018-10-29T08:44:42Z" level=error msg="/moby.buildkit.v1.Control/Solve returned error: exit code: 2
github.com/moby/buildkit/executor/runcexecutor.(*runcExecutor).Exec
	/go/src/github.com/moby/buildkit/executor/runcexecutor/executor.go:270
github.com/moby/buildkit/solver/llbsolver/ops.(*execOp).Exec
	/go/src/github.com/moby/buildkit/solver/llbsolver/ops/exec.go:645
github.com/moby/buildkit/solver.(*sharedOp).Exec.func1
	/go/src/github.com/moby/buildkit/solver/jobs.go:684
github.com/moby/buildkit/util/flightcontrol.(*call).run
	/go/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:116
github.com/moby/buildkit/util/flightcontrol.(*call).run-fm
	/go/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:142
sync.(*Once).Do
	/usr/local/go/src/sync/once.go:44
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1333
executor failed running [copy /src-0 go/src/handler/]
github.com/moby/buildkit/solver/llbsolver/ops.(*execOp).Exec
	/go/src/github.com/moby/buildkit/solver/llbsolver/ops/exec.go:646
github.com/moby/buildkit/solver.(*sharedOp).Exec.func1
	/go/src/github.com/moby/buildkit/solver/jobs.go:684
github.com/moby/buildkit/util/flightcontrol.(*call).run
	/go/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:116
github.com/moby/buildkit/util/flightcontrol.(*call).run-fm
	/go/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:142
sync.(*Once).Do
	/usr/local/go/src/sync/once.go:44
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1333"
```

**Comment by WTFKr0** [comment]
For information:

```
grep sda9 /proc/mounts
/dev/sda9 / ext4 rw,seclabel,relatime,data=ordered 0 0
```

```
cat /proc/fs/ext4/sda9/options
rw
block_validity
delalloc
barrier
user_xattr
acl
resuid=0
resgid=0
errors=remount-ro
commit=5
min_batch_time=0
max_batch_time=15000
stripe=0
data=ordered
inode_readahead_blks=32
init_itable=10
max_dir_size_kb=0
```

```
dumpe2fs /dev/sda9
Filesystem volume name:   ROOT
Last mounted on:          /sysroot
Filesystem UUID:          1eff66aa-15ea-4225-b11e-2d3f86d883c2
Filesystem magic number:  0xEF53
Filesystem revision #:    1 (dynamic)
Filesystem features:      has_journal ext_attr resize_inode dir_index filetype needs_recovery extent flex_bg sparse_super large_file huge_file uninit_bg dir_nlink extra_isize
Filesystem flags:         signed_directory_hash 
Default mount options:    user_xattr acl
Filesystem state:         clean
Errors behavior:          Remount read-only
Filesystem OS type:       Linux
Inode count:              7231872
Block count:              7257083
Reserved block count:     305636
Free blocks:              5161390
Free inodes:              7225518
First block:              0
Block size:               4096
Fragment size:            4096
Reserved GDT blocks:      134
Blocks per group:         32768
Fragments per group:      32768
Inodes per group:         32576
Inode blocks per group:   1018
Flex block group size:    16
Filesystem created:       Thu Sep 21 00:37:23 2017
Last mount time:          Fri Oct 19 09:33:30 2018
Last write time:          Fri Oct 19 09:33:30 2018
Mount count:              13
Maximum mount count:      -1
Last checked:             Thu Sep 21 01:25:23 2017
Check interval:           0 (<none>)
Lifetime writes:          3370 GB
Reserved blocks uid:      0 (user root)
Reserved blocks gid:      0 (group root)
First inode:              11
Inode size:	          128
Journal inode:            8
First orphan inode:       17998
Default directory hash:   half_md4
Directory Hash Seed:      4754c6d8-f251-4d3b-bde8-fc9339aefa6f
Journal backup:           inode blocks
Journal features:         journal_incompat_revoke
Journal size:             64M
Journal length:           16384
Journal sequence:         0x03a4c77d
Journal start:            9120
```

**Comment by alexellis** [comment]
OK so this isn't the XFS issue I was expecting and saw with @affix on CentOS, but it does appear to be due to your se_linux policies.

**Comment by AkihiroSuda** [comment]
What's kernel version? I suppose it is < 4.14, which lacks support for setting `security` xattr in userns?

**Comment by WTFKr0** [comment]
```
uname -r
4.14.48-coreos-r2
```

**Comment by WTFKr0** [comment]
This is the OS release : 
See https://coreos.com/releases/#1745.7.0
```
1745.7.0
Release Date: June 14, 2018
kernel: 4.14.48
rkt: 1.29.0
docker: 18.03.1
etcd: 3.3.3
systemd: 238
Ignition: 0.24.1
```

**Comment by WTFKr0** [comment]
By stracing the buildkit process, we see that this operation failed:
```
[pid  3917] lsetxattr("go/src/handler/Dockerfile", "security.selinux", "system_u:object_r:svirt_lxc_file_t:s0", 38, 0) = -1 EOPNOTSUPP (Operation not supported)
```

It's maybe a selinux issue, I have installed a default policy from coreos
This machine can build with no problem with the classic `docker build` command

**Comment by AkihiroSuda** [comment]
PR: https://github.com/containerd/continuity/pull/138

**Comment by AkihiroSuda** [comment]
Question: is this rootless mode or not?

**Comment by WTFKr0** [comment]
I've test with : 
- moby/builkit:v0.3.0
- moby/builkit:v0.3.1
- moby/builkit:master

So no rootless mode i think

**Comment by ivanayov** [comment]
@WTFKr0 can you confirm if build works with @AkihiroSuda's fix?
Seems that the problem is out of the OpenFaaS Cloud scope.

**Comment by WTFKr0** [comment]
No i don't
I don't even know what the fix of @AkihiroSuda do
It's seems to impact containerd lib, so to test that, i'll need to rebuild all the things
@AkihiroSuda how can i test your PR if it's ready ?

**Comment by AkihiroSuda** [comment]
Will share image soon

**Comment by AkihiroSuda** [comment]
could you try `akihirosuda/tmp-public:buildkit-copy-16@sha256:d6134c7b1ab7e1d4ad875c787d086f9ff7baae9eff3a26513cc6a8ee131a073b` (built from akihirosuda/buildkit_poc@4ca515dfe16fb0694e24d483914c81afc56fc95d)

**Comment by AkihiroSuda** [comment]
Confirmed the issue is reproducible and my patch solves the issue with [coreos-vagrant](https://github.com/coreos/coreos-vagrant) 1939.0.0, kernel `4.19.0-coreos #1 SMP Mon Oct 22 23:11:24 UTC 2018`

**Comment by WTFKr0** [comment]
OK seems to be ok with that image !
Thanx
Now i'm on a certificate problem on my private registry :smile:

**Comment by AkihiroSuda** [comment]
The patch is now available as `moby/buildkit:v0.3.2`

May I remove `akihirosuda/tmp-public:buildkit-copy-16`?

**Comment by WTFKr0** [comment]
Yeah i'm the only one who use it i think
I change my env to moby/buildkit:v0.3.2
